sync/server.md

```diff
@@ -0,0 +1,404 @@
+It will have the following parameters.
+
+* Operation e.g. find, insert, update, remove
+* Query (Only for some operations. See the nedb API)
+* Document (Only for some operations. See the nedb API)
+* Options (Only for some operations. See the nedb API)
+
+For example, to find a single record by id:
+
+```
+{
+  "operation": "find",
+  "query": {
+    "id": "{{context.arguments.id}}"
+  }
+}
+```
+
+Like all other request mappings, this will be treated as a string for Handlebars processing before being parsed as JSON.
+
+To create a record, here's an example:
+
+```
+{
+  "operation": "insert",
+  "doc": {
+    "title": "{{context.arguments.title}}",
+    "content": "{{context.arguments.content}}"
+  }
+}
+```
+
+
+### Resolver Response Mapping
+
+For In Memory responses, the `context.result` object will be result object passed back from the corresponding nedb API.
+For example, for an `insert`, the created document is returned. This can be mapped as follows:
+
+```
+{{toJSON context.result}}
+```
+
+However, for a `remove` operation, the nedb API returns a `numRemoved` value. This could be mapped to an Int in the Data Schema.
+
+```
+// removeNote(id: String) Int
+{
+  "numRemoved": {{context.result}}
+}
+```
+
+## Metrics
+
+The primary goals of Data Sync metrics are:
+
+* seeing resource usage of the Data Sync service over time
+* summary metrics about the number of queries/mutations resolved by the server
+* getting an insight into what Clients are connected/executing queries over time, the number of queries/mutations, which queries/mutations, and the size of data
+  * Clients connected for Subscriptions, over websocket, may be a good way of determining when a client is connected and condition of their network
+* being able to get Client specific information for debugging purposes e.g. a count of pending mutations that have built up while offline (depends on an offline mutation queueing feature)
+
+The above is not intended to limit what metrics to capture, but more as a starting point to discover what metrics are valueable.
+
+### Client Metrics
+
+There are 2 potential approaches to getting Client Metrics from a device to the Metrics service. The first approach is to use the MetricsService module in the AeroGear SDK to send metrics directly to the Metrics service. This would mean the Sync module in the SDK would need to call the Metrics service every so often.
+This is not a great approach as the device may not be online all the time.
+
+The second approach is to piggyback on any GraphQL requests, adding metrics data to the request context. The data will need to be limited in size to avoid unnecessary network overhead. Using simple counters that are maintained on the Client will help keep the payload size small. For example, the Sync module in the SDK maintains a specific set of counters as the App interacts with it.
+Whenever a Sync request (query/mutation) is made to the server, the latest values for this set of counters is sent along.
```

Comment:
Is this similar to  the "__fh" thing we did in the old fh js sdk? If so I don't think it is a good idea. If the whole request is defined by the developers, then I don't think we should change it while it's being sent from the clients.

Can we not reusing the existing metrics module and service?

Reply:
After off-thread discussions, using the metrics module in the client SDK makes sense if it is improved to add 2 features:

* custom metrics
* offline caching of metrics